Generate json view create sql scripts

ZdravDimOracle opened the pull request:
This pr adds functionality of generating create SQL scripts for creating JSON duality views based on Java source files.
For every class annotated with `@JsonView`, corresponding SQL for creating that view will be generated.
Fixes #3599.

ZdravDimOracle committed:
generate sql create json view statement

ZdravDimOracle committed:
generate sql create json view statement

ZdravDimOracle committed:
resolve merge conflict

renovate[bot] committed:
chore(deps): update plugin io.micronaut.build.shared.settings to v8.0.0-m12 (#3630)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/micronaut-projects/micronaut-data?pullRequest=3632) <br/>All committers have signed the CLA.

radovanradic commented on data-model/src/main/java/io/micronaut/data/model/PersistentEntity.java:
```diff
@@ -437,6 +438,26 @@ default PersistentPropertyPath getPropertyPath(@NonNull String[] propertyPath) {
+     * Get JSON view persistent entity
+     * @return The persistent entity
+     */
+    default Optional<PersistentEntity> getJsonViewEntity() {
```
If these are used only for creating SQL script then maybe not needed here and can be checked during create view build?

ZdravDimOracle replied:
These are used only for creating SQL script, but they call `getEntity()` function, which is protected in `RuntimePersistentEntity` class.

radovanradic commented on doc-examples/jdbc-example-java/src/main/java/example/ContactView.java:
```diff
@@ -0,0 +1,64 @@
+package io.micronaut.data.tck.entities;
```
package should be example and some classes are also missing here
Why are classed copied here if there are no tests using them?

ZdravDimOracle replied:
Tests inside `jdbc-example-java` module use `H2`, we can either change it to `Oracle` or reuse `ContactView` class from `data-tck`. Which approach to use?

radovanradic replied:
Don't see how is `ContactView` used in tests in jdbc-example-java. Looks like this class is not needed there at all, or I am missing something.

ZdravDimOracle replied:
There are no tests there, we wanted to update the documentation and provide some tests in `jdbc-example-java`. We will reuse the class from `data-tck` then.

radovanradic commented:
There are checkstyle errors

radovanradic commented on data-jdbc/src/main/java/io/micronaut/data/jdbc/config/SchemaGenerator.java:
```diff
@@ -270,6 +268,7 @@ private static void validate(Connection connection,
         // that represents join and ad-hoc SqlTableMapping for the same entity based on relation mappings (to be removed/skipped)
         Map<String, SqlTableMapping> sqlTableMappingByTableName = CollectionUtils.newLinkedHashMap(entities.length);
         for (PersistentEntity entity : entities) {
+            if (entity.getAnnotationMetadata().hasAnnotation(JsonView.class)) continue;
```
Should not skip if create view is supported now (for Oracle). If the issue is ContactView entity in data-tck being built for other dialects then can copy/move it where it's needed.

radovanradic replied:
Ok, this is for validation, it's fine.

graemerocher commented on data-model/src/main/java/io/micronaut/data/annotation/JsonSubView.java:
```diff
@@ -0,0 +1,88 @@
+import java.lang.annotation.Target;
+
+/**
+ * The annotation defining Json Duality SubView. Currently supported only by Oracle database.
```
need to elaborate this javadoc more. Describe what a subview is how it relates to the duality view feature in Oracle. Provide example code snippets etc.

graemerocher commented on data-model/src/main/java/io/micronaut/data/annotation/JsonSubView.java:
```diff
@@ -0,0 +1,88 @@
+     *
+     * @return the entity class (default void)
+     */
+    Class<?> entity() default void.class;
```
describe more in the javadoc what is a valid entity class

graemerocher commented on data-model/src/main/java/io/micronaut/data/annotation/JsonView.java:
```diff
@@ -41,6 +41,22 @@
+     */
+    Class<?> entity() default void.class;
+
+    enum Operation {UPDATE, INSERT, DELETE};
```
javadoc this and all enum members.

graemerocher commented on data-model/src/main/java/io/micronaut/data/model/query/builder/sql/SqlQueryBuilder.java:
```diff
@@ -25,11 +25,7 @@
-import io.micronaut.data.annotation.Join;
-import io.micronaut.data.annotation.Relation;
-import io.micronaut.data.annotation.Repository;
+import io.micronaut.data.annotation.*;
```
don't use star imports

graemerocher commented on data-model/src/main/java/io/micronaut/data/model/PersistentEntity.java:
```diff
@@ -437,6 +438,26 @@ default PersistentPropertyPath getPropertyPath(@NonNull String[] propertyPath) {
+     * Get JSON view persistent entity
+     * @return The persistent entity
+     */
+    default Optional<PersistentEntity> getJsonViewEntity() {
```
I would prefer to not add new methods to this interface and have a `JsonViewPersistentEntity` sub interface where these methods exists if possible. 

If this is not possible then extract these getters into a new `JsonViewInfo` type and return that from `PersistentEntity.viewInfo()`.

Also make sure each new public method is documented.

renovate[bot] committed:
chore(deps): update graalvm/setup-graalvm action to v1.4.4 (#3631)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

ZdravDimOracle committed:
remove old approach

ZdravDimOracle committed:
fix style check

ZdravDimOracle committed:
pr iteration

ZdravDimOracle committed:
Remove star import

ZdravDimOracle committed:
Remove star import

ZdravDimOracle committed:
update documentation

ZdravDimOracle committed:
update documentation

ZdravDimOracle committed:
Polish out code

andriy-dmytruk commented on data-jdbc/src/test/resources/application-oracle-jsonview.yml:
```diff
@@ -2,17 +2,9 @@ micronaut:
-        - classpath:oracle-jsonview-migrations
-      enabled: true
-      baseline-version: 0
-      baseline-on-migrate: true
```
Since we remove this, we should remove the `migrations/V1__Create_Schema.sql` and the flyway dependency

andriy-dmytruk commented on data-model/src/main/java/io/micronaut/data/annotation/JsonSubView.java:
```diff
@@ -0,0 +1,107 @@
+
+    /**
+     * The entity class.
+     * User can specify which entity class this JsonSubView corresponds to.
```
```suggestion
     * Specify an entity class annotated with {@link MappedEntity} that this JSON sub view corresponds to.
```

andriy-dmytruk commented on data-model/src/main/java/io/micronaut/data/annotation/JsonView.java:
```diff
@@ -41,6 +56,38 @@
+    /**
+     * The entity class.
+     *
+     * @return the entity class (default void)
```
Let's extend the javadoc here, too.

andriy-dmytruk commented on doc-examples/jdbc-example-java/src/main/java/example/ContactView.java:
```diff
@@ -0,0 +1,22 @@
+    LocalDateTime startDateTime,
+    boolean active
+) {}
+// end::record-example[]
```
Could we add a test here for inserting/querying this JSON view?

ZdravDimOracle committed:
remove flyway

ZdravDimOracle committed:
log error instead of throwing

radovanradic commented on data-jdbc/src/main/java/io/micronaut/data/jdbc/config/SchemaGenerator.java:
```diff
@@ -136,7 +138,8 @@ public void createOrValidateSchema(BeanLocator beanLocator) {
                 .filter(i -> !i.getBeanType().getName().contains("$"))
                 .filter(i -> !Modifier.isAbstract(i.getBeanType().getModifiers()))
-                .filter(i -> !i.hasAnnotation(JsonView.class))
+                .filter(i -> !i.hasAnnotation(JsonSubView.class))
```
Might need to change this in `R2dbcSchemaGenerator` as well and update test in data-r2dbc there is JsonView test with Oracle R2DBC.

ZdravDimOracle committed:
add r2dbc jsonview support

radovanradic commented on data-model/src/main/java/io/micronaut/data/model/query/builder/sql/SqlQueryBuilder.java:
```diff
@@ -365,9 +389,147 @@ public String[] buildCreateTableStatements(PersistentEntity[] entities) {
+        PersistentEntity entity = entityOptional.get();
+        String viewName = viewEntity.getPersistedName();
+        StringBuilder sb = new StringBuilder("CREATE OR REPLACE JSON RELATIONAL DUALITY VIEW ")
+            .append(viewName)
```
Maybe should check if there is schema configured.

radovanradic commented on data-model/src/main/java/io/micronaut/data/model/JsonViewPersistentEntity.java:
```diff
@@ -0,0 +1,47 @@
+
+import java.util.Optional;
+
+public interface JsonViewPersistentEntity {
```
Is this interface is needed, since it is called only from the `SqlQueryBuilder` then can just call from there
```
private Optional<PersistentEntity> getJsonView(@NonNull PersistentEntity entity) {
        if (entity.getAnnotationMetadata().hasAnnotation(JsonView.class)) {
            return entity.getAnnotationMetadata().classValue(JsonView.class, "entity").map(c -> PersistentEntity.of(c));
        }
        return Optional.empty();
    }
```
same for `JsonSubView` and not add changes to `PersistentEntity` etc.

radovanradic commented on data-model/src/main/java/io/micronaut/data/model/query/builder/sql/AbstractSqlLikeQueryBuilder.java:
```diff
@@ -366,14 +372,16 @@ protected String getPathOnlyAliasName(JoinPath joinPath) {
                              String currentJoinAlias,
-                             String lastJoinAlias) {
+                             String lastJoinAlias,
+                             boolean buildForJsonView) {
```
Can it be recognized if building for json view by presence of json view/sub view annotation in the entity? Then might avoid this parameter which makes sense only in `SqlQueryBuilder` and there check it and instead of calling 
```
private void join(StringBuilder builder,
                      @Nullable BaseQueryDefinition queryDefinition,
                      String joinType,
                      String tableName,
                      String tableAlias,
                      String onTableName,
                      List<String> onLeftColumns,
                      List<String> onRightColumns,
                      boolean buildForJsonView) {
```
just call this 
```
        builder.append(WHERE_CLAUSE);
        buildJoinColumnMatchPart(builder, tableAlias, onTableName, onLeftColumns, onRightColumns)
```
which is being executed currently if json view. Sorry if I missing something, but it would be better to avoid this specific param for all query builders, if possible.

ZdravDimOracle replied:
Yes, I think we can just check if associated entity has annotation `JsonSubView,` and continue based on that

ZdravDimOracle replied:
Actually, it's used for joining two entity classes, ex. `Student - Address` (not `StudentView` or `AddressSubView`), and they do not have the `@JsonView` or `@JsonSubView` annotations, so I think it's we can't remove this parameter.

radovanradic replied:
Maybe you can change implementation and provide QueryState with entity information when using JsonView for building query
```
QueryState queryState = new QueryState(new BaseQueryDefinition() {
            @Override
            public PersistentEntity persistentEntity() {
                return associatedViewEntity;
            }

            @Override
            public Predicate predicate() {
                return null;
            }

            @Override
            public Collection<JoinPath> getJoinPaths() {
                return List.of();
            }

            @Override
            public Optional<JoinPath> getJoinPath(String path) {
                return Optional.empty();
            }
        }, true, true);
```
in `SqlQueryBuilder.createJsonSubViewQuery` before call 
```
buildJoin(null, sb, queryState, joinAssociation, associationOwnerEntity, associatedEntity.getAliasName(), associationOwnerEntity.getAliasName(), true);
```
and then have information in queryState whether entity is JsonView or not. Then might revert nullability for QueryState and BaseQueryDefinition parameter changes.

ZdravDimOracle committed:
merge conflict and pr iteration

ZdravDimOracle committed:
forgot to revert

ZdravDimOracle committed:
check schema configured

ZdravDimOracle committed:
remove buildForJsonView parameter

radovanradic commented on data-model/src/main/java/io/micronaut/data/annotation/JsonView.java:
```diff
@@ -41,6 +56,41 @@
+     *
+     * @return the entity class (default void)
+     */
+    Class<?> entity() default void.class;
```
Is this required field, what happens if not configured for the JsonView? If it is, then should not default to Void and might add notes to the breaks.adoc since it is new field.

ZdravDimOracle replied:
It is required, if it's not set, `SqlQueryBuilder` will skip that entity. I changed it to not have a default value, and I updated the docs.

ZdravDimOracle committed:
remove buildForJsonView and flyway

ZdravDimOracle committed:
remove default entity value

ZdravDimOracle committed:
revert Address.java

ZdravDimOracle committed:
remove address methods

radovanradic commented on data-model/src/main/java/io/micronaut/data/model/query/builder/sql/AbstractSqlLikeQueryBuilder.java:
```diff
@@ -1376,7 +1382,10 @@ protected final void appendPropertyRef(AnnotationMetadata annotationMetadata,
-                        query.append('"').append(property.getPersistedName()).append('"');
+                            String persistedName = property.getAnnotationMetadata().stringValue(SERDE_CONFIG_ANNOTATION, "property")
+                                .orElse(property.getAnnotationMetadata().stringValue(JSON_PROPERTY_ANNOTATION)
+                                    .orElse(property.getName()));
```
shouldn't be `property.getPersistedName()` ?

radovanradic commented on data-model/src/main/java/io/micronaut/data/model/query/builder/sql/SqlQueryBuilder.java:
```diff
@@ -235,11 +232,16 @@ public String buildBatchDropTableStatement(@NonNull PersistentEntity... entities
     public String[] buildDropTableStatements(@NonNull PersistentEntity entity) {
+        List<String> dropStatements = new ArrayList<>();
+        if (entity.getAnnotationMetadata().hasAnnotation(JsonView.class)) {
+            String sql = "DROP " + getTableName(entity);
```
Should be `"DROP VIEW " + getTableName(entity)`

radovanradic commented on data-model/src/main/java/io/micronaut/data/model/query/builder/sql/SqlQueryBuilder.java:
```diff
@@ -1146,7 +1359,7 @@ protected void buildJoin(String joinType,
     private void join(StringBuilder sb,
                       String joinType,
-                      QueryState queryState,
+                      @Nullable QueryState queryState,
```
Can remove `@Nullable`

radovanradic commented on data-r2dbc/src/test/resources/application-oracle-jsonview.yaml:
```diff
@@ -6,20 +6,12 @@ datasources:
   default:
     db-type: oracle
     dialect: ORACLE
-flyway:
```
Can remove flyway from build.gradle maybe as not used by other tests

ZdravDimOracle committed:
remove flyway from build.gradle

graemerocher commented on data-jdbc/src/test/groovy/io/micronaut/data/jdbc/oraclexe/jsonview/Student.java:
```diff
@@ -11,15 +12,17 @@
 import java.util.List;
 
-@MappedEntity("TBL_STUDENT")
+@MappedEntity(value = "TBL_STUDENT", alias = "s")
```
is specifying an alias required or just for testing?

ZdravDimOracle replied:
It's just for testing purposes.

graemerocher commented on src/main/docs/guide/dbc/sqlMapping/sqlJsonView.adoc:
```diff
@@ -1,17 +1,17 @@
+Also, entity can be mapped to a JSON SUBVIEW as follows:
+
+snippet::example.AddressSubView[project-base="doc-examples/jdbc-example", source="main", tags="record-example"]
+
```
need more documentation on the following topics:

1. how to specify joins in the views (looks like this is done with JoinColumn)
2. how to use mapped property to map to the underlying column
3. how schema generation works
4. What is a JSON subview vs a JSON view and how do they differ
5. How to support different operations (update, delete etc.)

radovanradic replied:
Also, since now entity is required field on `JsonView` and is used only for creation - maybe can be changed not to be required and if not present then will log warn/error and skip creation of json view. Not sure if it makes sense, but I assume there are users already using JsonView without schema creation and might not know how to populate required value.

ZdravDimOracle committed:
update documentation

ZdravDimOracle committed:
merge conflict

ZdravDimOracle committed:
remove star imports

ZdravDimOracle committed:
remove star imports

graemerocher commented on src/main/docs/guide/dbc/sqlMapping/sqlJsonView.adoc:
```diff
@@ -1,17 +1,74 @@
+snippet::example.AddressSubView[project-base="doc-examples/jdbc-example", source="main", tags="record-example"]
+
+The main difference between `@JsonView` and `@JsonSubView` and the reason of existence of `@JsonSubView` annotation is to differentiate which for which class will
+creation sql script be generated (`JsonViews` are generated, while `JsonSubViews` are not).
```
there are some grammar errors here, run this through AI to get it corrected

graemerocher commented on src/main/docs/guide/dbc/sqlMapping/sqlJsonView.adoc:
```diff
@@ -1,17 +1,74 @@
+annotate your view class using `JoinColumn`, where `name` is the identifier for `StudentView`, and `referencedColumnName`
+is the identifier for `StudentScheduleSubView`, as follows:
+
+[source,java]
```
use `snippet::` and include a real code snippet

graemerocher commented on src/main/docs/guide/dbc/sqlMapping/sqlJsonView.adoc:
```diff
@@ -1,17 +1,74 @@
+You can define relationships between entities within JSON views by using appropriate annotations.
+For example, to associate a `StudentView` with its related `StudentScheduleSubView` entries in a one-to-many relationship,
+annotate your view class using `JoinColumn`, where `name` is the identifier for `StudentView`, and `referencedColumnName`
+is the identifier for `StudentScheduleSubView`, as follows:
```
can this `@JoinColumn` be inferred automatically? What happens if I don't specify it? Is there a compilation error?

ZdravDimOracle replied:
If you don't specify it, there is no compilation error, just the entity gets skipped. I think it can be inferred automatically, based on `@Id` annotation.

ZdravDimOracle replied:
It actually already works, but I didn't match property names for `StudentView` and `Student`. The property in `StudentView` needs to have `@JoinColumn` annotation, or it needs to match the property name in `Student`. I'm going to check what happens for `EmbeddedId` now.

ZdravDimOracle replied:
I pushed the changes, it works when class uses `@EmbeddedId` annotation, but joins currently don't work, so I'm working on it now.

ZdravDimOracle replied:
Just finished the work on joins with composite ids.

graemerocher commented on src/main/docs/guide/dbc/sqlMapping/sqlJsonView.adoc:
```diff
@@ -1,17 +1,74 @@
+==== Using `@MappedProperty` annotation
+
+Json view field names do not need to match the underlying table field names, and if it's not matching, then users must annotate the field
+using `@MappedProperty` annotation, where `value` field, represents the underlying table field name.
```
```suggestion
The fields of types annotated annotated with `@JsonView` or `@JsonSubview` do not need the match the corresponding entities field names as specified by the `entity` member (in the below case `Teacher`).

The `@MappedProperty` annotation can be used to control which underlying field the view matches in the entity.

The following example uses a different name in the view the `Teacher` entities `id` and `name` fields, remapping them to `teachID` and `teacher` respectively.

Json view field names do not need to match the underlying table field names, and if it's not matching, then users must annotate the field
```

graemerocher commented on src/main/docs/guide/dbc/sqlMapping/sqlJsonView.adoc:
```diff
@@ -1,17 +1,74 @@
+Json view field names do not need to match the underlying table field names, and if it's not matching, then users must annotate the field
+using `@MappedProperty` annotation, where `value` field, represents the underlying table field name.
+
+[source,java]
```
use `snippet::` and include an actual tested code snippet

graemerocher commented on src/main/docs/guide/dbc/sqlMapping/sqlJsonView.adoc:
```diff
@@ -1,17 +1,74 @@
+
+==== Operations supported
+
+Users can specify sql operations supported for `JsonView` and `JsonSubView` using the `operations` field, if not specified, it defaults to `{ INSERT, UPDATE, DELETE }`.
```
link to the oracle documentation on this topic somewhere

ZdravDimOracle replied:
Can I use this link?
https://docs.oracle.com/en/database/oracle/oracle-database/26/jsnvu/using-json-relational-duality-views.html

graemerocher commented on src/main/docs/guide/dbc/sqlMapping/sqlJsonView.adoc:
```diff
@@ -1,17 +1,74 @@
+
+Users can specify sql operations supported for `JsonView` and `JsonSubView` using the `operations` field, if not specified, it defaults to `{ INSERT, UPDATE, DELETE }`.
+
+[source,java]
```
use the `snippet::` macro and include actual code

ZdravDimOracle committed:
add imports

graemerocher commented on src/main/docs/guide/dbc/sqlMapping/sqlJsonView.adoc:
```diff
@@ -1,17 +1,74 @@
+
+==== Schema generation
+
+`JsonViews` are generated automatically for classes with `@JsonView` annotation.
```
this doesn't say anything. What I expect is something like 

```
Given the e example entities above the following JSON Duality View definition is generated:

ADD DUALITY VIEW GENERATED
```

You should use callouts like `<1>` , `<2>` to explain the duality view and link to the oracle docs on the subject

ZdravDimOracle replied:
I've added this example, but I didn't link to oracle docs, I'm not sure what link to use.

graemerocher replied:
https://docs.oracle.com/en/database/oracle/oracle-database/26/jsnvu/overview-json-relational-duality-views.html

graemerocher commented on src/main/docs/guide/dbc/sqlMapping/sqlJsonView.adoc:
```diff
@@ -1,17 +1,74 @@
 Essentially, json view will be treated like mapped entity and will return JSON structure from the database and be mapped to java entity. All CRUD operations can be
 performed against json view mapped entities.
 
+Also, entity can be mapped to a JSON SUBVIEW as follows:
```
here you refer to `SUBVIEW` and later `@JsonSubView`. Which is it?

Also when mentioning `@JsonView` and `@JsonSubview` you should use the `ann::` macro to link to the actual annotation in the javadoc. For example `ann:foo.bar.JsonView[]` (replace real package name)

ZdravDimOracle committed:
Update src/main/docs/guide/dbc/sqlMapping/sqlJsonView.adoc

Co-authored-by: Graeme Rocher <graeme.rocher@gmail.com>

ZdravDimOracle committed:
update documentation

ZdravDimOracle committed:
fix many to many joins

ZdravDimOracle committed:
fix composite ids

ZdravDimOracle committed:
Merge branch '5.0.x' of https://github.com/micronaut-projects/micronaut-data into Dimitrije/json-view-sql

ZdravDimOracle committed:
support json view creation for composite ids

ZdravDimOracle committed:
resolve merge conflict

ZdravDimOracle committed:
fix @Nullable errors

ZdravDimOracle committed:
fix checkstyle errors

ZdravDimOracle committed:
fix checkstyle errors

ZdravDimOracle committed:
fix join table primary keys

ZdravDimOracle committed:
fix join table primary keys

ZdravDimOracle committed:
update docs

radovanradic commented on data-jdbc/src/test/groovy/io/micronaut/data/jdbc/oraclexe/jsonview/OracleJdbcJsonViewSpec.groovy:
```diff
@@ -424,6 +427,17 @@ class OracleJdbcJsonViewSpec extends Specification {
+        apartments.add(ap1)
+        def buildingView = new BuildingView(123, apartments)
+        then:
+        buildingViewRepository.save(buildingView)
```
Can you add check for findById 
```
when:
        def buildingViewLoaded = buildingViewRepository.findById(buildingView.id()).orElse(null)
        then:
        buildingViewLoaded
        buildingViewLoaded.id() == buildingView.id()
```
and verify apartment data. Also, should not `AppartmentId` have building id (123) and not (1)

radovanradic commented on data-model/src/main/java/io/micronaut/data/model/query/builder/sql/SqlQueryBuilder.java:
```diff
@@ -512,12 +522,46 @@ private void createJsonViewQuery(StringBuilder sb, PersistentEntity viewEntity,
+                            sb.append(FROM_CLAUSE)
+                                .append(joinTableName)
+                                .append(SPACE)
+                                .append("WITH UPDATE INSERT DELETE");
```
Should it check for `JsonView.Operation[] operations()` instead of defaulting `UPDATE INSERT DELETE` ?

ZdravDimOracle replied:
Those operations are for the join table which gets created automatically without user action, the users cannot specify it, so it's using the default values.

andriy-dmytruk commented on data-model/src/main/java/io/micronaut/data/model/query/builder/sql/SqlQueryBuilder.java:
```diff
@@ -607,7 +891,8 @@ private String addGeneratedStatementToColumn(GeneratedValue.Type type, DataType
                     if (isPk) {
-                        column += " GENERATED ALWAYS AS IDENTITY (MINVALUE 1 START WITH 1 CACHE 100 NOCYCLE)";
+//                        column += " GENERATED ALWAYS AS IDENTITY (MINVALUE 1 START WITH 1 CACHE 100 NOCYCLE)";
+                        column += " GENERATED BY DEFAULT ON NULL AS IDENTITY (MINVALUE 1 START WITH 1 CACHE 100 NOCYCLE)";
```
Pleas remove the comment

ZdravDimOracle committed:
fix buildingViewTest and remove star imports

andriy-dmytruk commented on data-model/src/main/java/io/micronaut/data/model/query/builder/sql/SqlQueryBuilder.java:
```diff
@@ -357,9 +387,263 @@ public String[] buildCreateTableStatements(PersistentEntity[] entities) {
+                Field[] fields = ((RuntimePersistentProperty<?>) identity).getType().getDeclaredFields();
+                List<SqlColumnMapping> columnMappings = SqlSchemaUtils.getSqlTableMappings(entity).getFirst().primaryKeyColumns();
+                if (fields.length != columnMappings.size()) {
+                    return;
```
Should this throw an exception instead of returning?

andriy-dmytruk commented on data-model/src/main/java/io/micronaut/data/model/query/builder/sql/SqlQueryBuilder.java:
```diff
@@ -357,9 +387,263 @@ public String[] buildCreateTableStatements(PersistentEntity[] entities) {
+            if (identity.getAnnotationMetadata().hasAnnotation(MappedProperty.class)) {
+                entityPersistedPropertyName = identity.getPersistedName();
+            } else if (identity.getAnnotationMetadata().hasAnnotation(EmbeddedId.class)) {
+                if (isTopLevel) {
```
Could the embedded Id processing be separated into a separate method?

andriy-dmytruk commented on data-model/src/main/java/io/micronaut/data/model/query/builder/sql/SqlQueryBuilder.java:
```diff
@@ -357,9 +387,263 @@ public String[] buildCreateTableStatements(PersistentEntity[] entities) {
+                Relation.Kind kind = association.getKind();
+                Optional<PersistentEntity> associatedEntityOptional = getJsonSubViewEntity(association.getAssociatedEntity());
+                if (associatedEntityOptional.isEmpty()) {
+                    return;
```
Should this throw an exception instead?

andriy-dmytruk commented on data-model/src/main/java/io/micronaut/data/model/query/builder/sql/SqlQueryBuilder.java:
```diff
@@ -357,9 +387,263 @@ public String[] buildCreateTableStatements(PersistentEntity[] entities) {
+            String columnPropertyName = column.getAnnotationMetadata().stringValue(SERDE_CONFIG_ANNOTATION, "property")
+                .orElse(column.getAnnotationMetadata().stringValue(JSON_PROPERTY_ANNOTATION)
+                    .orElse(column.getName()));
+            if (column instanceof Association association) {
```
Could the association processing be separated into a separate method?

ZdravDimOracle committed:
throw exceptions instead of just returning

andriy-dmytruk commented on data-model/src/main/java/io/micronaut/data/model/query/builder/sql/SqlQueryBuilder.java:
```diff
@@ -640,6 +552,102 @@ private String createJsonSubViewQuery(Association association) {
         return sb.toString();
     }
 
+    private void processEmbeddedId(StringBuilder sb, PersistentEntity entity, PersistentProperty identity, String alias, Iterator<PersistentProperty> it, List<PersistentProperty> columns, boolean isTopLevel) {
```
Could you make the naming more complete, like also mention json schema, as those methods cannot be reused elsewhere.

ZdravDimOracle committed:
fix function naming

graemerocher commented on data-processor/src/test/groovy/io/micronaut/data/processor/sql/JsonViewSpec.groovy:
```diff
@@ -77,9 +64,6 @@ interface ContactViewRepository extends CrudRepository<ContactView, Long> {
         findAllOrderByStartDateTimeQuery == 'SELECT cv.* FROM "CONTACT_VIEW" cv ORDER BY cv.DATA.startDateTime.timestamp() ASC'
-        findByAddressStreetQuery == 'SELECT cv.* FROM "CONTACT_VIEW" cv WHERE (cv.DATA.address.street.stringOnly() = ?)'
-        findAddressStreetByIdQuery == 'SELECT cv.DATA.address.street.stringOnly() FROM "CONTACT_VIEW" cv WHERE (cv.DATA."_id".numberOnly() = ?)'
-        findAllOrderByAddressZipCodeDescQuery == 'SELECT cv.* FROM "CONTACT_VIEW" cv ORDER BY cv.DATA.address.zipCode.stringOnly() DESC'
```
can you explain why these assertions were deleted?

ZdravDimOracle replied:
I reverted these changes.

graemerocher commented on src/main/docs/guide/dbc/sqlMapping/sqlJsonView.adoc:
```diff
@@ -1,17 +1,87 @@
+
+NOTE: If ann:io.micronaut.data.annotation.JoinColumn[] is not specified, property name of Json View class needs to match the name of the corresponding entity class property name.
+
+NOTE: Using many to many relations in Json Duality View classes is not supported.
```
can we fail with a compilation error if the user attempts this?

ZdravDimOracle replied:
The doc is not up to date; I will update it now. You can check this [file](https://github.com/micronaut-projects/micronaut-data/pull/3632/changes#diff-e2f5a49cc213fba765e6463bfaf860170fd6e2048273d9e2218791eba84a06a9). it has many to many relation, so we already support many to many relations.

graemerocher commented on src/main/docs/guide/dbc/sqlMapping/sqlJsonView.adoc:
```diff
@@ -1,17 +1,87 @@
+joined automatically, on columns annotated with ann:io.micronaut.data.annotation.Id[] by default, but if user specified columns to join in
+by ann:io.micronaut.data.annotation.JoinColumn[], those columns will be used.
+
+NOTE: If ann:io.micronaut.data.annotation.JoinColumn[] is not specified, property name of Json View class needs to match the name of the corresponding entity class property name.
```
is this compilation time validated?

ZdravDimOracle replied:
No, this currently happens in runtime.

ZdravDimOracle committed:
add compile time checks

ZdravDimOracle committed:
merge

ZdravDimOracle committed:
remove flyway

ZdravDimOracle committed:
fix checkstyle

andriy-dmytruk committed:
Improve MappedEntityVisitor changes to resolve the build errors

ZdravDimOracle committed:
Merge branch '5.0.x' of https://github.com/micronaut-projects/micronaut-data into Dimitrije/json-view-sql

ZdravDimOracle committed:
merge branch '5.0.x' of https://github.com/micronaut-projects/micronaut-data into Dimitrije/json-view-sql

ZdravDimOracle committed:
Merge branch '5.0.x' of https://github.com/micronaut-projects/micronaut-data into Dimitrije/json-view-sql

ZdravDimOracle committed:
Merge branch 'Dimitrije/json-view-sql' of https://github.com/ZdravDimOracle/micronaut-data into Dimitrije/json-view-sql

ZdravDimOracle committed:
remove unsused address property

ZdravDimOracle committed:
remove unsused import

ZdravDimOracle committed:
Add address subview

andriy-dmytruk committed:
Merge branch 'refs/heads/5.0.x' into Dimitrije/json-view-sql

andriy-dmytruk committed:
Revert to use embedded entity address

andriy-dmytruk committed:
Fix handling of Kind.Embedded associations with JsonViews

ZdravDimOracle committed:
add table_name field for @JsonView annotation

andriy-dmytruk committed:
Merge remote-tracking branch 'ZdravDim/Dimitrije/json-view-sql' into Dimitrije/json-view-sql

ZdravDimOracle committed:
revert table_name

andriy-dmytruk committed:
Do not add delete operation for StudentSubView as it cannot be deleted recursively

andriy-dmytruk committed:
Merge branch '5.0.x' into Dimitrije/json-view-sql

andriy-dmytruk committed:
Resolve failures

andriy-dmytruk committed:
Add Relation.unique property to mark relations that should not be repeated

- Modify SqlSchemaUtils to create primary key in this case
- Revert setting all columns as primary key by default. The default for unique is false

andriy-dmytruk commented on data-model/src/main/java/io/micronaut/data/annotation/Relation.java:
```diff
@@ -44,6 +44,14 @@
+     * For example, if true, a primary key might be created in the mapping table to enforce uniqueness.
+     * @return Whether the relation is unique.
+     */
+    boolean unique() default false;
```
I added this property to relation, which marks that a relation cannot be repeated. We needed it, as json view `UNNEST` command required the mapping table to have a primary key.

andriy-dmytruk replied:
As discussed with Radovan, I will revert this and instead we will always make the mapping table columns as primary key. I will mark the failing test with PendingIssue referencing https://github.com/micronaut-projects/micronaut-data/issues/3722.

andriy-dmytruk committed:
Revert "Add Relation.unique property to mark relations that should not be repeated"

This reverts commit 11e7fdb33d4e6190d16dbf90c522a9c6e8ce6074.

andriy-dmytruk committed:
Mark failing test with pending feature and explanation
